components/raftstore/src/store/peer.rs

```diff
@@ -4795,6 +4795,7 @@ where
             let read_ts = decode_u64(&mut req.get_header().get_flag_data()).unwrap();
             let safe_ts = self.read_progress.safe_ts();
             if safe_ts < read_ts {
+                self.read_progress.notify_advance_resolved_ts();
```

Comment:
It can be expansive for every read request to lock.

Reply:
It's not for every read, but for rejected read.

Reply:
Once safe ts is not advanced, rejection should be the major branches.

Reply:
In practice, there won't be many requests reach the branch, because the latency is high (in my tests, ~4 seconds) it will block clients from sending further requests. Anyway, I will address your comments.